core/lib/zksync_core/src/consensus/mod.rs

```diff
@@ -61,61 +60,155 @@ impl MainNodeConfig {
-#[derive(Debug, Clone)]
-pub struct FetcherConfig {
-    pub executor: executor::Config,
-}
+    /// Fetches blocks from the main node in range `[cursor.next()..end)`.
+    async fn fetch_blocks(
+        &self,
+        ctx: &ctx::Ctx,
+        cursor: &mut storage::Cursor,
+        end: validator::BlockNumber,
+    ) -> ctx::Result<()> {
+        const MAX_CONCURRENT_REQUESTS: usize = 100;
+        let begin = cursor.next();
+        scope::run!(ctx, |ctx, s| async {
+            let (send, mut recv) = ctx::channel::bounded(MAX_CONCURRENT_REQUESTS);
+            s.spawn(async {
+                let send = send;
+                let sub = &mut self.sync_state.subscribe();
+                for n in begin.0..end.0 {
+                    let n = MiniblockNumber(n.try_into().unwrap());
+                    sync::wait_for(ctx, sub, |state| state.main_node_block() >= n).await?;
+                    send.send(ctx, s.spawn(self.fetch_block(ctx, n))).await?;
+                }
+                Ok(())
+            });
+            while cursor.next() < end {
+                let block = recv.recv(ctx).await?.join(ctx).await?;
+                cursor.advance(block).await?;
+            }
+            Ok(())
+        })
+        .await
+    }
 
-impl FetcherConfig {
     /// Task fetching L2 blocks using peer-to-peer gossip network.
     pub async fn run(
         self,
         ctx: &ctx::Ctx,
-        pool: ConnectionPool,
+        store: Store,
         actions: ActionQueueSender,
     ) -> anyhow::Result<()> {
-        scope::run!(ctx, |ctx, s| async {
-            let store = Store::new(pool);
-            let mut block_store = store.clone().into_block_store();
-            block_store
-                .set_actions_queue(ctx, actions)
+        let res: ctx::Result<()> = scope::run!(ctx, |ctx, s| async {
+            // Update sync state in the background.
+            s.spawn_bg(self.fetch_state_loop(ctx));
+
+            // Initialize genesis.
+            let genesis = self.fetch_genesis(ctx).await.wrap("fetch_genesis()")?;
+            let mut conn = store.access(ctx).await.wrap("access()")?;
+            conn.try_update_genesis(ctx, &genesis)
                 .await
-                .wrap("block_store.set_actions_queue()")?;
+                .wrap("set_genesis()")?;
+            let mut cursor = conn
+                .new_fetcher_cursor(ctx, actions)
+                .await
+                .wrap("new_fetcher_cursor()")?;
+            drop(conn);
+
+            // Fetch blocks before the genesis.
+            self.fetch_blocks(ctx, &mut cursor, genesis.fork.first_block)
+                .await?;
+
+            // Monitor the genesis of the main node.
+            // If it changes, it means that a hard fork occurred and we need to reset the consensus state.
+            s.spawn_bg::<()>(async {
+                let want = genesis;
+                loop {
+                    if let Ok(got) = self.fetch_genesis(ctx).await {
```

Comment:
Checking genesis could be the scope of responsibility of the reorg detector, or maybe another small component ("consensus reorg detector"? "fork detector"?) that would work similarly to the reorg detector. I don't have a particularly strong opinion regarding this, so feel free to ignore.

Reply:
This "fork detector" is implemented as a subtask of `Fetcher::run()`. I don't see a reason to flatten the task structure here and expose "fork detector" to the outside logic. Fork detector needs to start only once the consensus genesis for this execution is fixed, so it is not like it can be independent from `Fetcher::run()` execution.